### PR TITLE
feat(routesv2/trafficsplit): Traffic split for routes v2

### PR DESCRIPTION
**Pull request by snehachhabria**
**Description**:

This PR has the final set of changes to integrate traffic split with
routes v2. The changes in the PR address issues #2545 and #1979.

This PR further incorporates the routes v2 feature flag as a part of our
automated demo, giving the users the ability to enable/disable routes v2
based on the value set in their .env file.

Signed-off-by: Sneha Chhabria <snchh@microsoft.com>

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [X]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [X]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution? `no`

### CODE REVIEW

File: pkg/catalog/types.go [hunk 1]
```diff
@@ -143,9 +143,11 @@ type MeshCataloger interface {
...
 	ListInboundTrafficTargetsWithRoutes(service.K8sServiceAccount) ([]trafficpolicy.TrafficTargetWithRoutes, error)
 
-	// GetApexServicesForBackend returns a list of services that serve as the apex service in a traffic split where the
-	//	given service is a backend
-	GetApexServicesForBackend(service.MeshService) []service.MeshService
+	// IsApexService returns true if the given service is an apex service in any traffic split
+	IsApexService(service.MeshService) bool
+
+	// GetBackendServiceForApexService returns a list of backend services that serve the given apex service in a traffic split
+	GetBackendServiceForApexService(apexSvc service.MeshService) []service.MeshService
```
Comment (shashankram) [inline pkg/catalog/types.go:150]: This is internal to the `MeshCatalog` implementation, no reason to expose it in the interface.

Comment (snehachhabria) [reply pkg/catalog/types.go]: good catch! I have updated the code

File: pkg/catalog/routes.go [hunk 2]
```diff
@@ -556,12 +556,23 @@ func (mc *MeshCatalog) buildInboundPolicies(t *access.TrafficTarget) []*trafficp
...
-
-		for _, sourceServiceAccount := range trafficTargetIdentitiesToSvcAccounts(t.Spec.Sources) {
-			for _, routeMatch := range routeMatches {
-				servicePolicy.AddRule(*trafficpolicy.NewRouteWeightedCluster(routeMatch, weightedCluster), sourceServiceAccount)
+		// while building inbound policies, check if the service is an apex service in traffic split
+		if mc.isApexService(destService) {
+			var localDestServices []service.MeshService
+			// build the inbound policy for an apex service to have its weighted clusters point to the backend services
+			backendServices := mc.getBackendServiceForApexService(destService)
+			// among all the backedn services, select only those that operlap with the destination services
```
Comment (shashankram) [inline pkg/catalog/routes.go:564]: typo: s/backedn/backend, s/operlap/overlap

File: pkg/catalog/routes.go [hunk 3]
```diff
@@ -556,12 +556,23 @@ func (mc *MeshCatalog) buildInboundPolicies(t *access.TrafficTarget) []*trafficp
...
+			// (i.e the destination service account)
```
Comment (shashankram) [inline pkg/catalog/routes.go:565]: `destination services (i.e. destination service account)` is confusing. Maybe rephrase.

Comment (snehachhabria) [reply pkg/catalog/routes.go]: removed `(i.e. destination service account)` . Hopefully that reduces the confusion

**Review by draychev — APPROVED** [review]
looks good to me - excited to start testing this

File: demo/deploy-bookstore.sh [hunk 4]
```diff
@@ -25,10 +23,7 @@ spec:
   ports:
   - port: 80
     name: bookstore-port
-  selector:
```
[the remark this answers is not in the document]
Comment (snehachhabria) [reply demo/deploy-bookstore.sh:28]: we don't need the selector on services that are used as apex services in a traffic split, that's the reason we got rid of it.

**Review by shashankram — APPROVED** [review]
Thanks!

File: pkg/catalog/routes.go [hunk 5]
```diff
@@ -572,6 +582,18 @@ func (mc *MeshCatalog) buildInboundPolicies(t *access.TrafficTarget) []*trafficp
 	return inboundPolicies
 }
 
+func buildInboundPolicyRules(servicePolicy *trafficpolicy.InboundTrafficPolicy, routeMatches []trafficpolicy.HTTPRouteMatch, sourceSvcAccounts []service.K8sServiceAccount, services []service.MeshService) *trafficpolicy.InboundTrafficPolicy {
```
Comment (michelleN) [inline pkg/catalog/routes.go:585]: What do you think about moving this func to pkg/trafficpolicy and hanging this as a method on *trafficpolicy.InboundTrafficPolicy?

File: pkg/catalog/routes.go [hunk 6]
```diff
@@ -556,12 +556,22 @@ func (mc *MeshCatalog) buildInboundPolicies(t *access.TrafficTarget) []*trafficp
...
 		servicePolicy := trafficpolicy.NewInboundTrafficPolicy(buildPolicyName(destService, false), hostnames)
 
-		weightedCluster := getDefaultWeightedClusterForService(destService)
-
-		for _, sourceServiceAccount := range trafficTargetIdentitiesToSvcAccounts(t.Spec.Sources) {
-			for _, routeMatch := range routeMatches {
-				servicePolicy.AddRule(*trafficpolicy.NewRouteWeightedCluster(routeMatch, weightedCluster), sourceServiceAccount)
+		// while building inbound policies, check if the service is an apex service in traffic split
+		if mc.isApexService(destService) {
+			var localDestServices []service.MeshService
```
Comment (michelleN) [inline pkg/catalog/routes.go:561]: Could you explain why we're checking if the service is an apex on inbound?

**Review by michelleN — CHANGES_REQUESTED** [review]

File: pkg/catalog/routes.go (hunk 5 above)
Comment (snehachhabria) [reply pkg/catalog/routes.go:585]: I did think of that but `getDefaultWeightedClusterForService` is component of catalog and will not be available to use in pkg/trafficpolicy. Further catalog cannot be imported in trafficpolicy as it will cause circular dependency

File: pkg/catalog/routes.go (hunk 6 above)
Comment (snehachhabria) [reply pkg/catalog/routes.go:561]: In the case of building inbound routes for a particular service we need the check if its an apex service (with traffic split) cause the hostnames need to be the name of the apex service but its weighted cluster needs to be the backend service.

Comment (snehachhabria) [reply pkg/catalog/routes.go:561]: We previously had it as the apex service itself but that didn't work cause we don't fully have support for multiple services on a single pod. This will require more work

**Review by michelleN — APPROVED** [review]
